File: Sources/custom_adapter/AdmobAdapter/Objc/AlxAdmobInterstitialAdapter.m
//
//  AlxAdmobInterstitialAdapter.m
//  AlxAdsOCDemo
//

#import "AlxAdmobInterstitialAdapter.h"
#import <AlxAds/AlxAds-Swift.h>

static NSString *const TAG = @"AlxAdmobInterstitialAdapter";

@interface AlxAdmobInterstitialAdapter () <AlxInterstitialAdDelegate>

@property (nonatomic, strong, nullable) AlxInterstitialAd *interstitialAd;
@property (nonatomic, weak, nullable) id<GADMediationInterstitialAdEventDelegate> delegate;
@property (nonatomic, copy, nullable) GADMediationInterstitialLoadCompletionHandler completionHandler;

@end

@implementation AlxAdmobInterstitialAdapter

- (void)loadInterstitialForAdConfiguration:(GADMediationInterstitialAdConfiguration *)adConfiguration
                         completionHandler:(GADMediationInterstitialLoadCompletionHandler)completionHandler {
    NSLog(@"%@: loadInterstitial", TAG);
    
    NSDictionary *params = [AlxAdmobBaseAdapter parseAdparameterFor:adConfiguration.credentials];
    if (!params) {
        NSString *errorStr = @"The parameter field is not found in the adConfiguration object";
        NSLog(@"%@: config params is empty", TAG);
        self.delegate = completionHandler(nil, [self errorWithCode:-100 msg:errorStr]);
        return;
    }
    
    if (!AlxAdmobBaseAdapter.isInitialized) {
        [AlxAdmobBaseAdapter initSdkFor:params];
    }
    
    NSString *adId = params[@"unitid"];
    if (!adId || adId.length == 0) {
        NSString *errorStr = @"unitid is empty in the parameter configuration";
        NSLog(@"%@: error: %@", TAG, errorStr);
        self.delegate = completionHandler(nil, [self errorWithCode:-100 msg:errorStr]);
        return;
    }
    
    NSLog(@"%@: loadInterstitial unitid=%@", TAG, adId);
    self.completionHandler = completionHandler;
    
    // Load ad
    self.interstitialAd = [[AlxInterstitialAd alloc] init];
    self.interstitialAd.delegate = self;
    [self.interstitialAd loadAdWithAdUnitId:adId];
}

- (void)presentFromViewController:(UIViewController *)viewController {
    NSLog(@"%@: present", TAG);
    if (self.interstitialAd && [self.interstitialAd isReady]) {
        [self.interstitialAd showAdWithPresent:viewController];
    }
}

#pragma mark - AlxInterstitialAdDelegate

- (void)interstitialAdLoad:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdLoad", TAG);
    if (self.completionHandler) {
        self.delegate = self.completionHandler(self, nil);
    }
}

- (void)interstitialAdFailToLoad:(AlxInterstitialAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: interstitialAdFailToLoad: %@", TAG, error.localizedDescription);
    if (self.completionHandler) {
        self.delegate = self.completionHandler(nil, error);
    }
}

- (void)interstitialAdImpression:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdImpression", TAG);
    [self.delegate willPresentFullScreenView];
    [self.delegate reportImpression];
}

- (void)interstitialAdClick:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdClick", TAG);
    [self.delegate reportClick];
}

- (void)interstitialAdClose:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdClose", TAG);
    [self.delegate didDismissFullScreenView];
}

- (void)interstitialAdRenderFail:(AlxInterstitialAd *)ad didFailWithError:(NSError *)error {
    NSLog(@"%@: interstitialAdRenderFail", TAG);
}

- (void)interstitialAdVideoStart:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdVideoStart", TAG);
}

- (void)interstitialAdVideoEnd:(AlxInterstitialAd *)ad {
    NSLog(@"%@: interstitialAdVideoEnd", TAG);
}

@end

File: custom_adapter/AdmobAdapter/Swift/AlxAdmobInterstitialAdapter.swift
//
//  AlxAdmobInterstitialAdapter.swift
//

import Foundation
import GoogleMobileAds
import AlxAds

@objc(AlxAdmobInterstitialAdapter)
public class AlxAdmobInterstitialAdapter: AlxAdmobBaseAdapter,MediationInterstitialAd {
    
    private static let TAG = "AlxAdmobInterstitialAdapter"
    
    private var interstitialAd: AlxInterstitialAd? = nil
    private var delegate: MediationInterstitialAdEventDelegate? = nil
    private var completionHandler: GADMediationInterstitialLoadCompletionHandler? = nil
    
    public func loadInterstitial(for adConfiguration: MediationInterstitialAdConfiguration, completionHandler: @escaping GADMediationInterstitialLoadCompletionHandler) {
        NSLog("%@: loadInterstitial",AlxAdmobInterstitialAdapter.TAG)
        guard let params = AlxAdmobBaseAdapter.parseAdparameter(for: adConfiguration.credentials) else {
            let errorStr="The parameter field is not found in the adConfiguration object"
            NSLog("%@: config params is empty",AlxAdmobInterstitialAdapter.TAG)
            self.delegate=completionHandler(nil,self.error(code: -100,msg: errorStr))
            return
        }
        
        if !AlxAdmobBaseAdapter.isInitialized {
            AlxAdmobBaseAdapter.initSdk(for: params)
        }
        
        guard let adId = params["unitid"] as? String,!adId.isEmpty else{
            let errorStr="unitid is empty in the parameter configuration"
            NSLog("%@: error: %@",AlxAdmobInterstitialAdapter.TAG,errorStr)
            self.delegate=completionHandler(nil,self.error(code: -100,msg: errorStr))
            return
        }
        
        NSLog("%@: loadInterstitial unitid=%@",AlxAdmobInterstitialAdapter.TAG,adId)
        self.completionHandler = completionHandler
        
        // load ad
        self.interstitialAd = AlxInterstitialAd()
        self.interstitialAd?.delegate = self
        self.interstitialAd?.loadAd(adUnitId: adId)
    }
    
    public func present(from viewController: UIViewController) {
        NSLog("%@: present",AlxAdmobInterstitialAdapter.TAG)
        
        if let interstitialAd = self.interstitialAd,interstitialAd.isReady(){
            interstitialAd.showAd(present: viewController)
        }
    }
    
}

extension AlxAdmobInterstitialAdapter:AlxInterstitialAdDelegate {
    
    
    public func interstitialAdLoad(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdLoad",AlxAdmobInterstitialAdapter.TAG)
        if let handler=self.completionHandler{
            self.delegate=handler(self,nil)
        }
    }
    
    public func interstitialAdFailToLoad(_ ad: AlxInterstitialAd, didFailWithError error: Error) {
        NSLog("%@: interstitialAdFailToLoad : %@",AlxAdmobInterstitialAdapter.TAG,error.localizedDescription)
        if let handler=self.completionHandler{
            self.delegate=handler(nil,error)
        }
    }
    
    public func interstitialAdImpression(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdImpression",AlxAdmobInterstitialAdapter.TAG)
        self.delegate?.willPresentFullScreenView()
        self.delegate?.reportImpression()
    }
    
    public func interstitialAdClick(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdClick",AlxAdmobInterstitialAdapter.TAG)
        self.delegate?.reportClick()
    }
    
    public func interstitialAdClose(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdClose",AlxAdmobInterstitialAdapter.TAG)
        self.delegate?.didDismissFullScreenView()
    }
    
    public func interstitialAdRenderFail(_ ad: AlxInterstitialAd, didFailWithError error: Error) {
        NSLog("%@: interstitialAdRenderFail",AlxAdmobInterstitialAdapter.TAG)
    }
    
    public func interstitialAdVideoStart(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdVideoStart",AlxAdmobInterstitialAdapter.TAG)
    }
    
    public func interstitialAdVideoEnd(_ ad: AlxInterstitialAd) {
        NSLog("%@: interstitialAdVideoEnd",AlxAdmobInterstitialAdapter.TAG)
    }
    
}
